### ISSUE TITLE
Statically type check "generic" methods 

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/dev_compiler#300]
We need to be able to statically type check methods such as Iterable.map.  E.g., this should not trigger a warning:

```dart
Iterable<int> incr(Iterable<int> iter) => iter.map((int x) => x + 1);
```

Initially, we can annotate / whitelist methods that should be treated as generic.

See #28 for general bug on generic methods.

**Comment by vsmenon** [comment]
Note, the code in the first comment is now handled by our initial whitelist implementation, but we're not doing any inference yet.  The following variant (with less typing on the closure) is still not handled:

```dart
Iterable<int> incr(Iterable<int> iter) => iter.map((x) => x + 1);
```

**Comment by alorenzen** [comment]
Another example of "generic" methods is Future.then, in which a Future<T>.then((T)->V) should return a Future<V>.

**Comment by jmesserly** [comment]
Vijay fixed this a while ago (at first with a hard coded list). Support for comment-based annotations is mostly done as well.
